### PR TITLE
Bug 1263233 - Issues in guided dtable analysis - ConditionInspectors

### PR DESCRIPTION
**Pull request by Rikkola**
After quite a long time and quite a few headaches they all pass.
Added few extra tests, just to split the problems into smaller bits.

This covers everything in the PR here. https://github.com/droolsjbpm/drools-wb/pull/64

### CODE REVIEW

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/condition/StringConditionInspector.java [hunk 1]
```diff
@@ -76,62 +81,217 @@ public boolean conflicts( Object other ) {
...
+                                return !otherInspector.containsAll( values );
                         }
                     case NOT_EQUALS:
-                        return !( (StringConditionInspector) other ).values.contains( values.get( 0 ) );
+                        switch (otherInspector.getOperator()) {
+                            case IN:
+                                return doesNotContainAll( ((StringConditionInspector) other).getValues() );
+                            case NOT_EQUALS:
+                                return !otherInspector.containsAll( values );
+//                            case EQUALS:
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/condition/StringConditionInspector.java]: Want to leave this commented out?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/condition/StringConditionInspector.java [hunk 2]
```diff
@@ -76,62 +81,217 @@ public boolean conflicts( Object other ) {
...
+                                  final Operator b ) {
+        return (this.operator.equals( a ) && otherInspector.getOperator().equals( b ))
+                || (this.operator.equals( b ) && otherInspector.getOperator().equals( a ));
+    }
+
+    @Override
+    public boolean covers( Comparable<String> otherValue ) {
+
+        switch (operator) {
+            //TODO: waaayyyy too simple
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/condition/StringConditionInspector.java]: Is this PR incomplete?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/checks/util/OperatorTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    @Test
+    public void testOperators() throws Exception {
+        assertEquals( Operator.EQUALS, Operator.resolve( "==" ) );
+        assertEquals( Operator.GREATER_THAN, Operator.resolve( ">" ) );
+        assertEquals( Operator.LESS_THAN, Operator.resolve( "<" ) );
+        assertEquals( Operator.GREATER_THAN_OR_EQUALS, Operator.resolve( ">=" ) );
+        assertEquals( Operator.LESS_THAN_OR_EQUALS, Operator.resolve( "<=" ) );
+        assertEquals( Operator.NOT_EQUALS, Operator.resolve( "!=" ) );
+        assertEquals( Operator.IN, Operator.resolve( "in" ) );
+        assertEquals( Operator.NOT_IN, Operator.resolve( "not in" ) );
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/checks/util/OperatorTest.java]: There are a bunch more CEP operators than these 3 :(

**Comment by manstis** [comment]
Liking the parametrized tests :)

**Comment by triceo** [comment]
Thanks @zkrejcov  and @Rikkola, this is some awesome work! It would have been a nightmare, trying to test all of this in Selenium.

**Comment by Rikkola** [comment]
Need to clean this up. 
@manstis Yep, learned something new, works well for this scenario.
@triceo Really good work from @zkrejcov, pointing out what is wrong with unit tests. Then it is up to me to fix things.

**Comment by Rikkola** [comment]
I need a local git hook that screams at me if I commit "TODO".

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/condition/StringConditionInspector.java (hunk 1 above)
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/condition/StringConditionInspector.java]: Nope, I want it left out. We should not commit commented code.
Thanks.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/checks/util/OperatorTest.java (hunk 3 above)
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/checks/util/OperatorTest.java]: Looks like there are. Still need to update this.

**Comment by Rikkola** [comment]
Just updated this, but noticed I still need to check that we can handle memberOf operator.

**Comment by Rikkola** [comment]
Ok, nothing we can do with memberOf. At the moment at least. There are ways to figure out what might be in a collection, but the level that this V&V is now does not handle it.
PR is ready for a new review.

**Comment by manstis** [comment]
*drools-wb*
(master) https://github.com/droolsjbpm/drools-wb/commit/fa99c8a18

**Pull request closed without merging**